src/desktop/WebDialog.ts

```diff
@@ -0,0 +1,149 @@
+import {app, BrowserWindow, ipcMain, WebContents} from "electron"
+import path from "path"
+import {defer} from "@tutao/tutanota-utils"
+import {ElectronWebContentsTransport} from "./ipc/ElectronWebContentsTransport.js"
+import {NativeToWebRequest, WebToNativeRequest} from "../native/main/WebauthnNativeBridge.js"
+import {MessageDispatcher} from "../api/common/MessageDispatcher.js"
+import {exposeRemote} from "../api/common/WorkerProxy.js"
+import {CancelledError} from "../api/common/error/CancelledError.js"
+import {CentralIpcHandler} from "./ipc/CentralIpcHandler.js"
+import {register} from "./electron-localshortcut/LocalShortcut.js"
+
+/**
+ * a browserWindow wrapper that
+ * * opens a specific website
+ * * installs a nativeBrigde
+ * * sends a request to the webContents
+ * * returns the result of the call
+ */
+
+export interface IWebDialogController {
+	create<FacadeType extends object>(parentWindowId: number, urlToOpen: URL): Promise<WebDialog<FacadeType>>
+}
+
+export const webauthnIpcConfig = Object.freeze({
+	renderToMainEvent: "to-main-webdialog",
+	mainToRenderEvent: "to-renderer-webdialog"
+})
+
+export type WebDialogIpcConfig = typeof webauthnIpcConfig
+export type WebDialogIpcHandler = CentralIpcHandler<WebDialogIpcConfig>
+
+// Singleton
+export const webauthnIpcHandler: WebDialogIpcHandler = new CentralIpcHandler(ipcMain, webauthnIpcConfig)
+
+/** A dialog which was already loaded. Allows sending one request or closing it. */
+export class WebDialog<FacadeType extends object> {
+	constructor(
+		private facade: FacadeType,
+		private closedPromise: Promise<never>,
+		private browserWindow: BrowserWindow,
+	) {
+	}
+
+	makeRequest<T>(requestSender: (facade: FacadeType) => Promise<T>): Promise<T> {
+		return Promise
+			.race([
+				this.closedPromise,
```

Comment:
Is there a reason that `closedPromise` needs to throw an exception when the dialog is closed and not just return normally? It's not really an exceptional case.

Reply:
It doesn't return a result because it's cancelled. We don't really use monadic results in many places though.

Reply:
ok